File: scripts/bcra/itcrm/update_itcrm.js
const fs = require("fs");
const path = require("path");
const XLSX = require("xlsx");

const BASE_URL = "https://www.bcra.gob.ar/archivos/Pdfs/PublicacionesEstadisticas/ITCRMSerie.xlsx";

const BACKEND_ROOT = path.resolve(__dirname, "../../..");
const OUTPUT_DIR = path.join(BACKEND_ROOT, "src", "data", "itcrm");
const OUTPUT_DAILY = path.join(OUTPUT_DIR, "itcrm_diario.csv");
const OUTPUT_MONTHLY = path.join(OUTPUT_DIR, "itcrm_mensual.csv");

function asegurarCarpeta() {
	fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

async function descargarXlsx(url) {
	const response = await fetch(url);

	if (!response.ok) {
		throw new Error(`No se pudo descargar el Excel. Estado HTTP: ${response.status}`);
	}

	const arrayBuffer = await response.arrayBuffer();
	return Buffer.from(arrayBuffer);
}

function excelSerialToDate(valor) {
	if (valor === null || valor === undefined || valor === "") return null;

	if (valor instanceof Date) {
		return valor;
	}

	if (typeof valor === "number") {
		if (valor < 30000) return null;
		return new Date(Math.round((valor - 25569) * 86400 * 1000));
	}

	const parsed = Date.parse(String(valor));
	if (!Number.isNaN(parsed)) return new Date(parsed);

	return null;
}

function formatDateYYYYMMDD(date) {
	if (!(date instanceof Date) || Number.isNaN(date.getTime())) return null;

	const year = date.getFullYear();
	const month = String(date.getMonth() + 1).padStart(2, "0");
	const day = String(date.getDate()).padStart(2, "0");
	return `${year}-${month}-${day}`;
}

function csvEscape(valor) {
	const texto = String(valor ?? "");

	if (texto.includes(",") || texto.includes('"') || texto.includes("\n")) {
		return `"${texto.replace(/"/g, '""')}"`;
	}

	return texto;
}

function guardarCsv(rows, outFile, header) {
	const lines = rows.map((row) => header.map((key) => csvEscape(row[key])).join(","));
	fs.writeFileSync(outFile, [header.join(","), ...lines].join("\n"), "utf8");
}

function parseSheet(workbook, sheetName) {
	const sheet = workbook.Sheets[sheetName];
	const sheetData = XLSX.utils.sheet_to_json(sheet, { header: 1, defval: "" });

	const rows = [];

	for (let i = 2; i < sheetData.length; i += 1) {
		const fila = sheetData[i] || [];
		const fecha = formatDateYYYYMMDD(excelSerialToDate(fila[0]));
		const valor = Number(fila[1]);

		if (!fecha || !Number.isFinite(valor)) continue;

		rows.push({ fecha, valor: Number(valor.toFixed(12)) });
	}

	return rows;
}

async function main() {
	asegurarCarpeta();

	console.log("Descargando ITCRM desde BCRA...");
	console.log(`URL: ${BASE_URL}`);

	const buffer = await descargarXlsx(BASE_URL);
	console.log("Archivo descargado correctamente y parseado en memoria.");

	const workbook = XLSX.read(buffer, { type: "buffer" });

	const sheetDaily = workbook.SheetNames.find((name) => /itcrm y bilaterales$/i.test(name)) || workbook.SheetNames[0];
	const sheetMonthly = workbook.SheetNames.find((name) => /prom\. mens\./i.test(name)) || workbook.SheetNames[1] || workbook.SheetNames[0];

	const daily = parseSheet(workbook, sheetDaily);
	const monthly = parseSheet(workbook, sheetMonthly);

	// Completar meses faltantes en mensual usando promedios del diario
	const mesActual = new Date().toISOString().slice(0, 7);
	const mesesExistentes = new Set(monthly.map((r) => r.fecha.slice(0, 7)));

	const porMes = new Map();
	for (const d of daily) {
		const mes = d.fecha.slice(0, 7);
		if (!porMes.has(mes)) porMes.set(mes, { suma: 0, count: 0, ultimaFecha: d.fecha });
		const entry = porMes.get(mes);
		entry.suma += d.valor;
		entry.count += 1;
		entry.ultimaFecha = d.fecha;
	}

	const monthlyCompleto = [...monthly];
	for (const [mes, entry] of porMes) {
		if (mes >= mesActual) continue; // solo meses cerrados
		if (mesesExistentes.has(mes)) continue;
		const promedio = entry.suma / entry.count;
		monthlyCompleto.push({ fecha: entry.ultimaFecha, valor: Number(promedio.toFixed(12)) });
	}
	monthlyCompleto.sort((a, b) => a.fecha.localeCompare(b.fecha));

	guardarCsv(daily, OUTPUT_DAILY, ["fecha", "valor"]);
	guardarCsv(monthlyCompleto, OUTPUT_MONTHLY, ["fecha", "valor"]);

	const lastDaily = daily[daily.length - 1] || null;
	const lastMonthly = monthlyCompleto[monthlyCompleto.length - 1] || null;

	console.log(`Hoja diaria: ${sheetDaily}`);
	console.log(`Filas diarias generadas: ${daily.length}`);
	console.log(`Archivo generado: ${OUTPUT_DAILY}`);
	if (lastDaily) console.log(`Último dato diario: ${lastDaily.fecha} = ${lastDaily.valor}`);

	console.log(`Hoja mensual: ${sheetMonthly}`);
	console.log(`Filas mensuales base: ${monthly.length} | con completado diario: ${monthlyCompleto.length}`);
	console.log(`Archivo generado: ${OUTPUT_MONTHLY}`);
	if (lastMonthly) console.log(`Último dato mensual: ${lastMonthly.fecha} = ${lastMonthly.valor}`);
}

main().catch((error) => {
	console.error("Error actualizando ITCRM:");
	console.error(error.message);
	process.exit(1);
});

File: scripts/bcra/itcrm/update_itcrm.py
from __future__ import annotations

import csv
import subprocess
from pathlib import Path

import pandas as pd
import requests

BASE_URL = "https://www.bcra.gob.ar/archivos/Pdfs/PublicacionesEstadisticas/ITCRMSerie.xlsx"

BACKEND_ROOT = Path(__file__).resolve().parents[3]
OUTPUT_DIR = BACKEND_ROOT / "src" / "data" / "itcrm"
OUTPUT_DAILY = OUTPUT_DIR / "itcrm_diario.csv"
OUTPUT_MONTHLY = OUTPUT_DIR / "itcrm_mensual.csv"


def asegurar_carpeta() -> None:
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)


def descargar_xlsx(url: str) -> bytes:
    try:
        response = requests.get(url, timeout=90)
        response.raise_for_status()
        return response.content
    except requests.RequestException:
        result = subprocess.run(
            ["curl.exe", "--silent", "--show-error", "--location", "--fail", url],
            capture_output=True,
            check=True,
        )
        return result.stdout


def normalizar_fecha(valor) -> str | None:
    if pd.isna(valor):
        return None

    try:
        dt = pd.to_datetime(valor)
    except Exception:  # noqa: BLE001
        return None

    if pd.isna(dt):
        return None

    return dt.strftime("%Y-%m-%d")


def parse_sheet(excel_path: Path, sheet_name: str) -> list[dict]:
    raw = pd.read_excel(excel_path, sheet_name=sheet_name, header=None)
    rows: list[dict] = []

    for i in range(2, len(raw)):
        fecha = normalizar_fecha(raw.iloc[i, 0])
        valor_raw = raw.iloc[i, 1]
        try:
            valor = float(valor_raw)
        except (TypeError, ValueError):
            continue

        if not fecha:
            continue

        rows.append({"fecha": fecha, "valor": round(valor, 12)})

    rows.sort(key=lambda x: x["fecha"])
    return rows


def guardar_csv(rows: list[dict], output_file: Path) -> None:
    with output_file.open("w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=["fecha", "valor"])
        writer.writeheader()
        writer.writerows(rows)


def completar_mensual_con_diario(daily: list[dict], monthly: list[dict]) -> list[dict]:
    mes_actual = pd.Timestamp.now().strftime("%Y-%m")
    meses_existentes = {r["fecha"][:7] for r in monthly}

    por_mes: dict[str, dict] = {}
    for d in daily:
        mes = d["fecha"][:7]
        if mes not in por_mes:
            por_mes[mes] = {"suma": 0.0, "count": 0, "ultima_fecha": d["fecha"]}
        por_mes[mes]["suma"] += float(d["valor"])
        por_mes[mes]["count"] += 1
        por_mes[mes]["ultima_fecha"] = d["fecha"]

    monthly_completo = list(monthly)
    for mes, entry in por_mes.items():
        if mes >= mes_actual:
            continue
        if mes in meses_existentes:
            continue
        promedio = entry["suma"] / entry["count"]
        monthly_completo.append({"fecha": entry["ultima_fecha"], "valor": round(promedio, 12)})

    monthly_completo.sort(key=lambda x: x["fecha"])
    return monthly_completo


def main() -> None:
    asegurar_carpeta()

    print("Descargando ITCRM desde BCRA...")
    print(f"URL: {BASE_URL}")

    content = descargar_xlsx(BASE_URL)
    temp_file = OUTPUT_DIR / "_itcrm_tmp.xlsx"
    temp_file.write_bytes(content)
    print("Archivo descargado correctamente y parseado en memoria.")

    excel = pd.ExcelFile(temp_file)
    sheet_names = excel.sheet_names

    sheet_daily = next((n for n in sheet_names if n.lower().endswith("itcrm y bilaterales")), sheet_names[0])
    sheet_monthly = next((n for n in sheet_names if "prom. mens." in n.lower()), sheet_names[1] if len(sheet_names) > 1 else sheet_names[0])

    daily = parse_sheet(temp_file, sheet_daily)
    monthly = parse_sheet(temp_file, sheet_monthly)
    monthly_completo = completar_mensual_con_diario(daily, monthly)

    guardar_csv(daily, OUTPUT_DAILY)
    guardar_csv(monthly_completo, OUTPUT_MONTHLY)

    try:
        temp_file.unlink(missing_ok=True)
    except Exception:  # noqa: BLE001
        pass

    last_daily = daily[-1] if daily else None
    last_monthly = monthly_completo[-1] if monthly_completo else None

    print(f"Hoja diaria: {sheet_daily}")
    print(f"Filas diarias generadas: {len(daily)}")
    print(f"Archivo generado: {OUTPUT_DAILY}")
    if last_daily:
        print(f"Último dato diario: {last_daily['fecha']} = {last_daily['valor']}")

    print(f"Hoja mensual: {sheet_monthly}")
    print(f"Filas mensuales base: {len(monthly)} | con completado diario: {len(monthly_completo)}")
    print(f"Archivo generado: {OUTPUT_MONTHLY}")
    if last_monthly:
        print(f"Último dato mensual: {last_monthly['fecha']} = {last_monthly['valor']}")


if __name__ == "__main__":
    main()
